dictionary.go

```diff
@@ -296,3 +314,118 @@ func (a *adder) addWordUnknownWord(w string, countable bool) {
 		a.failed++
 	}
 }
+
+// a librarian collates dictionaries.
+type librarian struct {
+	rules map[string]string
+}
+
+// newLibrarian returns a new librarian populated with words and affix rules
+// obtained from the hunspell .dic file paths provided, checking that the
+// affix file aff also exists.
+func newLibrarian(aff, dic string) (librarian, error) {
+	_, err := os.Stat(aff)
+	if err != nil {
+		return librarian{}, err
+	}
+	l := librarian{make(map[string]string)}
+	err = l.addDictionary(dic)
+	if err != nil {
+		return librarian{}, err
+	}
+	return l, nil
+}
+
+// addDictionary adds word rules from the hunspell dictionary at the given
+// path.
+func (l librarian) addDictionary(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	sc := bufio.NewScanner(f)
+	for i := 0; sc.Scan(); i++ {
+		if i == 0 {
+			// Skip word count line.
+			continue
+		}
+		err := l.addWord(sc.Text())
+		if err != nil {
+			return fmt.Errorf("%w at %s:%d", err, path, i+1)
+		}
+	}
+	return sc.Err()
+}
+
+// addWord adds the provided word to the librarian's dictionary merging any
+// affix rules into those already existing for the word.
+func (l librarian) addWord(w string) error {
+	r := strings.Split(w, "/")
+	word := r[0]
+	if word == "" {
+		// This should never happen, but we can ignore it.
+		return nil
+	}
+	var affix string
+	switch len(r) {
+	case 1:
+	case 2:
+		affix = r[1]
+	default:
+		return fmt.Errorf("invalid dictionary entry %q", w)
+	}
+	l.rules[word] = mergeRules(l.rules[word], affix)
+	return nil
+}
+
+// mergeRules merges affix rules.
+func mergeRules(a, b string) string {
+	switch {
+	case a == "":
+		return b
+	case b == "":
+		return a
+	default:
+		r := make([]rune, 0, utf8.RuneCountInString(a)+utf8.RuneCountInString(b))
+		r = append(r, []rune(a)...)
+		r = append(r, []rune(b)...)
```

Comment:
the merging requires `[]rune`, and strings are immutable, so this kind of merging is pretty GC heavy. have you thought about `map[string][]rune` instead?

Reply:
an alternative is to keep the merging pretty simple, e.g. `map[string][]string`, and at the very end join all the strings into a single `[]rune`, sort, and deduplicate.

Reply:
yet another idea, since this seems to be a rune set: `map[string]map[rune]struct{}`

Reply:
Yeah. I wasn't particularly happy with this. Given that the rules are sorted after the first pass, checking for equality before doing a dedup would save a lot of work. I doubt the map approaches would be cheaper than what I have here which is a rune-set-by-ordering.

Reply:
Thinking more carefully about it, a map[string][]rune would halve the work here.